File: packages/signals/src/signals/owner-response-pattern.ts
import type { SignalResult } from '@signalyze/shared';
import thresholds from '../../data/thresholds.json';
import { ramp, round6 } from '../math';
import type { PreparedReview } from '../prepare';
import { unavailable } from '../types';

const T = thresholds.owner_response_pattern;

/**
 * Among reviews with an owner response, how identical the responses are:
 * identicalShare = 1 - distinct / responded. value = identicalShare.
 */
export function ownerResponsePattern(reviews: readonly PreparedReview[]): SignalResult {
  const responses: string[] = [];
  for (const r of reviews) {
    if (r.normalizedOwnerResponse !== null && r.normalizedOwnerResponse !== '') {
      responses.push(r.normalizedOwnerResponse);
    }
  }
  const responded = responses.length;
  const responseRate = responded / reviews.length;
  if (responded < T.minResponded) {
    return unavailable('owner_response_pattern', {
      responded,
      responseRate: round6(responseRate),
      minResponded: T.minResponded,
    });
  }
  const groups = new Map<string, number>();
  for (const text of responses) groups.set(text, (groups.get(text) ?? 0) + 1);
  let largest = 0;
  for (const c of groups.values()) if (c > largest) largest = c;
  const identicalShare = 1 - groups.size / responded;
  return {
    id: 'owner_response_pattern',
    unusualness: round6(ramp(identicalShare, T.low, T.high)),
    value: round6(identicalShare),
    details: {
      identicalShare: round6(identicalShare),
      responded,
      responseRate: round6(responseRate),
      distinctResponses: groups.size,
      largestGroupShare: round6(largest / responded),
    },
    available: true,
  };
}

File: apps/api/signalyze_api/engine/signals/owner_response_pattern.py
"""Twin of ``signals/owner-response-pattern.ts``."""

from __future__ import annotations

from collections.abc import Sequence

from signalyze_api.engine.data_files import THRESHOLDS
from signalyze_api.engine.mathutil import ramp, round6
from signalyze_api.engine.prepare import PreparedReview
from signalyze_api.engine.types import unavailable
from signalyze_api.models import SignalResult

T = THRESHOLDS.owner_response_pattern


def owner_response_pattern(reviews: Sequence[PreparedReview]) -> SignalResult:
    """Among reviews with an owner response, how identical the responses are.

    identicalShare = 1 - distinct / responded. value = identicalShare.
    """
    responses: list[str] = []
    for r in reviews:
        if r.normalized_owner_response is not None and r.normalized_owner_response != "":
            responses.append(r.normalized_owner_response)
    responded = len(responses)
    response_rate = responded / len(reviews)
    if responded < T.minResponded:
        return unavailable(
            "owner_response_pattern",
            {
                "responded": responded,
                "responseRate": round6(response_rate),
                "minResponded": T.minResponded,
            },
        )
    groups: dict[str, int] = {}
    for text in responses:
        groups[text] = groups.get(text, 0) + 1
    largest = 0
    for c in groups.values():
        if c > largest:
            largest = c
    identical_share = 1 - len(groups) / responded
    return SignalResult(
        id="owner_response_pattern",
        unusualness=round6(ramp(identical_share, T.low, T.high)),
        value=round6(identical_share),
        details={
            "identicalShare": round6(identical_share),
            "responded": responded,
            "responseRate": round6(response_rate),
            "distinctResponses": len(groups),
            "largestGroupShare": round6(largest / responded),
        },
        available=True,
    )
